x-pack/plugins/maps/public/layers/sources/xyz_tms_source/xyz_tms_editor.tsx

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { Fragment, Component, ChangeEvent } from 'react';
+import _ from 'lodash';
+import { EuiFormRow, EuiFieldText } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+
+interface AttributionUpdate {
+  attributionText?: string;
+  attributionUrl?: string;
+}
+export interface XYZTMSSourceConfig extends AttributionUpdate {
+  urlTemplate: string;
+}
+
+export interface Props {
+  onSourceConfigChange: (sourceConfig: XYZTMSSourceConfig) => void;
+}
+
+interface State {
+  tmsInput: string;
+  tmsCanPreview: boolean;
+  attributionText?: string;
+  attributionUrl?: string;
+}
+
+export class XYZTMSEditor extends Component<Props, State> {
+  state = {
+    tmsInput: '',
+    tmsCanPreview: false,
+    attributionText: '',
+    attributionUrl: '',
+  };
+
+  _sourceConfigChange = _.debounce((updatedSourceConfig: XYZTMSSourceConfig) => {
+    if (this.state.tmsCanPreview) {
+      this.props.onSourceConfigChange(updatedSourceConfig);
+    }
+  }, 2000);
+
+  _handleTMSInputChange(e: ChangeEvent<HTMLInputElement>) {
+    const url = e.target.value;
+
+    const canPreview =
+      url.indexOf('{x}') >= 0 && url.indexOf('{y}') >= 0 && url.indexOf('{z}') >= 0;
+    this.setState(
+      {
+        tmsInput: url,
+        tmsCanPreview: canPreview,
+      },
+      () => this._sourceConfigChange({ urlTemplate: url })
```

Comment:
There is actually a bug here. If you change the urlTemplate after setting attribution, then the attribution will not get passed to `_sourceConfigChange` and will not be applied. I think the fix for this is for `_sourceConfigChange` to take no parameters and then just have the implementation pull values from state.

Reply:
prefer to fix in different PR since this is existing issue. Ideally, the duplicated attribution code is consolidated too cf. https://github.com/elastic/kibana/issues/32821#issuecomment-610463241